### PR TITLE
Drive-by mention unsafe fn closure coercion

### PR DESCRIPTION
**Pull request by alercah**

### CODE REVIEW

**Comment by ehuss** [comment]
I'm a bit uncertain about why it would call out `unsafe` specifically here. `unsafe fn` isn't a distinct pointer type from `fn` is it?  I wouldn't expect to call out all the function "front matter" like "extern" here, either.

**Comment by alercah** [comment]
Yes, they are. Interestingly, though, `unsafe` closures can be assigned to regular function pointers: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=de1dee0081843286d774d1d21c27091c

(I'm not sure this is wrong, but it's interesting.)

I forgot about `extern`, but closures can't be converted to extern functions (other than usual Rust ones).

**Comment by ldm0** [comment]
A drive-by developer here.

@alercah There is no "unsafe closure" actually. For the `let u = || unsafe {};`, `u` is just a closure with an unsafe block.

But I think this PR maybe useful. Since the rustc currently support coercing `non-capturing closure` to [safe](https://github.com/rust-lang/rust/issues/39817) and [unsafe](https://github.com/rust-lang/rust/pull/59580) function pointer. Coercing `non-capturing closure` to `extern` function pointer may be implemented in the future(<https://github.com/rust-lang/rust/issues/44291>) but now it won't pass compilation. I guess the reference need to point out either the compile-passing `unsafe fn` or the non-compile-passing `extern fn`.

**Comment by nikomatsakis** [comment]
I do not think of `unsafe fn` as a distinct type from `fn` -- rather I think of "function pointers" as types that have a range of options, something like this (ignoring `for<...>`):

* `extern "ABI" unsafe? fn(A0..An) -> R`

**Comment by alercah** [comment]
What about impls?

**Comment by nikomatsakis** [comment]
@alercah 

> so perhaps we should just write "(e.g., fn())" instead.

what do you think about this? :)

**Comment by alercah** [comment]
Ah, I see. That makes sense to me!

File: src/types/closure.md [hunk 1]
```diff
@@ -131,8 +131,8 @@ more specific call traits:
 > not how it captures them.
 
 *Non-capturing closures* are closures that don't capture anything from their
-environment. They can be coerced to function pointers (`fn`) with the matching
-signature.
+environment. They can be coerced to function pointers (`fn` or `unsafe fn`)
```
Comment (nikomatsakis) [inline src/types/closure.md:134]:
```suggestion
environment. They can be coerced to function pointers (e.g., `fn()`)
```

**Comment by nikomatsakis** [comment]
@ehuss how do I merge, just click the green button? :)

**Comment by ehuss** [comment]
Yes, after marking "approved" just click the merge button. 😃

**Pull request merged**